Improvement of regex-interpolation trap section

threadless-screw opened the pull request:
## The problem

The section is unclear and seems to cause more confusion than it solves. For instance, phrases like "the latter only works sometimes" and  "$(…) lexically interpolates the given string. So <{…}> immediately breaks with more complicated inputs" provide no accurate description of anything.

## The solution

The distinction between `$(...)` and `<{...}>` is currently described in the (updated) regex interpolation section of regex.pod6, which makes it clear that for the second example in the traps section to work it should have the to-be-interpolated string defined with a quoted hash, like this: `my $x = 'ailemac"#"';`, simply because `<{...}>` interpolates the return value/string as a regex. Since the situation is now clearly described in the regexes doc, this section may be deleted.

AlexDaniel commented:
Just because something is described somewhere in the docs doesn't mean that people wouldn't erroneously use the wrong feature.

> which makes it clear that for the second example in the traps section to work it should have the to-be-interpolated string defined with a quoted hash

Correct, and the point is that often this is used with arbitrary input, so quoting something is not really a solution to anything. IMO people should prefer `$($x)` over `<{$x}>` in almost all cases, but this is currently not the case.

That section should stay, but we can clarify it if it's unclear.

threadless-screw commented:
> But trap documentation is created for a reason, and I don't think that reason has disappeared, so I would rather have it stay.

Can you clarify then what the reason is that this documentation was created? To me it seems the reason was that the difference between the two constructs `$(...)` and `<{...}` was not understood (likely because it was not clearly documented), a problem which has been dealt with in the updated regexes page. 

> ...and the point is that often this is used with arbitrary input, so quoting something is not really a solution to anything. IMO people should prefer $($x) over <{$x}> in almost all cases, but this is currently not the case.

Can you give an example of arbitrary input that causes a problem? Or indicate why people should use `$(...) over `<{...}>`, even though they do different things and one is not a straightforward alternative for the other? It seems to me that one should "prefer" '$(...)' in case literal interpolation is desired, and '<{...}>' in case interpolation as a regex is desired, in much the same way that people should "prefer" `+` if addition is desired, and '-' if subtraction is desired.

In summary: I don't understand the point that this section is trying to communicate. And it seems that what it _is_ communicating is confusing if not plain wrong. Even the main message in the heading `<{$x}> vs $($x): Implicit EVAL`, i.e. that `<{...}>` uses IMPLICIT EVAL while `$(...)` does not, does not seem correct; at any rate, both constructs can evaluate arbitrary code.

JJ commented:
if there are issues with the text, it's much better if you raise the issues
and try to improve what we have instead of deleting it.


El mié., 7 ago. 2019 a las 9:12, Alexander (<notifications@github.com>)
escribió:

> But trap documentation is created for a reason, and I don't think that
> reason has disappeared, so I would rather have it stay.
>
> Can you clarify then what the reason is that this documentation was
> created? To me it seems the reason was that the difference between the two
> constructs $(...) and <{...} was not understood (likely because it was
> not clearly documented), a problem which has been dealt with in the updated
> regexes page.
>
> ...and the point is that often this is used with arbitrary input, so
> quoting something is not really a solution to anything. IMO people should
> prefer $($x) over <{$x}> in almost all cases, but this is currently not the
> case.
>
> Can you give an example of arbitrary input that causes a problem? Or
> indicate why people should use $(...) over <{...}>, even though they do
> different things and one is not a straightforward alternative for the
> other? It seems to me that one should "prefer" '$(...)' in case literal
> interpolation is desired, and '<{...}>' in case interpolation as a regex is
> desired, in much the same way that people should "prefer" +` if addition
> is desired, and '-' if subtraction is desired.
>
> In summary: I don't understand the point that this section is trying to
> communicate. And it seems that what it *is* communicating is confusing if
> not plain wrong. Even the main message in the heading <{$x}> vs $($x):
> Implicit EVAL, i.e. that <{...}> uses IMPLICIT EVAL while $(...) does
> not, does not seem correct; at any rate, both constructs can evaluate
> arbitrary code.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/perl6/doc/pull/2940?email_source=notifications&email_token=AAAAD5GRGOITPQWJBODTTNLQDJYWHA5CNFSM4IJ2T7DKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3XOAKA#issuecomment-518971432>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAAD5BQY4RZGTZ47V6YDD3QDJYWHANCNFSM4IJ2T7DA>
> .
>


-- 
JJ

threadless-screw commented:
That's not answering the pivotal question: what is the point this section is trying to communicate? Without knowing what point needs to be communicated, there is no sense in trying to clarify the text.

AlexDaniel commented:
> Can you give an example of arbitrary input that causes a problem?

Let's say you want to match a flipped variable. Like `$x.flip`. Both `$($x.flip)` and `<{$x.flip}>` will work for simple cases, but not for all input (so it's not really like `+` and `-`). Example of input that exposes the problem is `$x = ‘ailemac#’`. Sometimes people use `<{}>` when they actually need `$()`. That sections tells not to do that.

AlexDaniel commented:
That being said, I like the idea of removing some stuff that is not useful. But that section is not that.

JJ commented:
El mié., 7 ago. 2019 a las 9:42, Aleks-Daniel Jakimenko-Aleksejev (<
notifications@github.com>) escribió:

> That being said, I like the idea of removing some stuff that is not
> useful. But that section is not that.
>

Also, the best place to discuss this kind of things are issues, not pull
requests. Ideally, a PR should address an issue.

threadless-screw commented:
> Let's say you want to match a flipped variable. Like $x.flip. Both $($x.flip) and <{$x.flip}> will work for simple cases, but not for all input (so it's not really like + and -). Example of input that exposes the problem is $x = ‘ailemac#’. Sometimes people use <{}> when they actually need $(). That sections tells not to do that.

I see. But the updated regex page (note: at the time of writing only to be found in the github repo, as it is not rendered under docs.perl6.org yet) essentially does the same, also in a side-by-side description/comparison of interpolation constructs.

Anyway, although I think it smells a bit of documentation bloat, I prepared a new/rewritten trap section that aims to retain the above pointer, but without the confusing languange.

AlexDaniel commented:
I'll review this today. There are some things that need to be changed.

JJ commented on doc/Language/traps.pod6:
```diff
@@ -921,46 +921,64 @@ of the assignment operators instead:
-/language/pragmas#index-entry-MONKEY-SEE-NO-EVAL-MONKEY-SEE-NO-EVAL> pragma).
-In theory, careless use of C«<{}>» will only result in an exception being
-thrown, and should not introduce security issues.
+=head1 Interpolation constructs
```
Maybe this should be `=head2`? Also, it might be better if it was formulated like an actual trap, something like "Problems in the interpretation of regex interpolation constructs"

threadless-screw replied:
`head1` was my bad. - As to the title, I picked smth short because the long titles wrap somewhat clumsily when rendered in HTML. But if we go for long, how about 'Interpolation constructs that behave similarly, some of the time' ?

JJ replied:
OK

JJ commented on doc/Language/traps.pod6:
```diff
@@ -921,46 +921,64 @@ of the assignment operators instead:
+
+Perl 6 offers several constructs to generate regexes at runtime through
+interpolation (see their detailed description
+L<here|/language/regexes#Regex_interpolation>). When a thus generated regex
```
maybe better "regex generated this way"?

threadless-screw replied:
ok

JJ commented on doc/Language/traps.pod6:
```diff
@@ -921,46 +921,64 @@ of the assignment operators instead:
+Perl 6 offers several constructs to generate regexes at runtime through
+interpolation (see their detailed description
+L<here|/language/regexes#Regex_interpolation>). When a thus generated regex
+contains only characters that match themselves, some of these constructs behave
```
"character that match themselves" == "literal characters"? We already know them match themselves.

threadless-screw replied:
The thing is that, somewhat surprisingly, the main regex page does not make a clear distinction between literal characters and metacharacters. The latter term does not even occur there at all, and so it is obviously not defined there either. Since I regard the use of undefined terminology as a deficiency (exactly the type that we're trying to fix (lexcially interpolates?)), I included a descriptive definition of metacharacters in this section.

The better option would probably be to first amend the main regex page to define the terms 'literal character' and 'metacharacter' properly, and then to rely on their definition here without an attempt to clarify them (again). If you agree, I'll make a concurrent proposal for such an amendment to the regex page.

threadless-screw replied:
I just prepared a parallel PR to properly introduce the term 'metacharacter' in the regexes page. Once accepted, this PR can be cleaned up a bit.

JJ commented on doc/Language/traps.pod6:
```diff
@@ -921,46 +921,64 @@ of the assignment operators instead:
+
+But when the variable is changed to comprise regex metacharacters, i.e.
+characters that are neither alphanumeric nor the underscore C<_>, the outputs
+become different:
```
Eliminate the explanation of metacharacters, not needed here.

threadless-screw replied:
I disagree. See above comment.

JJ replied:
Still not needed. You merged the metacharacters section already, point to there.

threadless-screw replied:
Indeed. Now we agree it should go :-) I’ll rework the text soon.

JJ replied:
This is the only outstanding issue, right?

threadless-screw replied:
I removed the explanation of metacharacters from the traps section earlier. I should be gone.

JJ replied:
You didn't link the new section, but fair enough. We can do that afterwards.

threadless-screw commented:
@AlexDaniel I noticed you self-assigned this PR. Does this mean you prefer to fix the issue with the regex-interpolation traps section yourself? Then I'll leave it for what it is and let it dangle...

AlexDaniel commented:
> @AlexDaniel I noticed you self-assigned this PR. Does this mean you prefer to fix the issue with the regex-interpolation traps section yourself? Then I'll leave it for what it is and let it dangle...

Not really, just wanted to make sure I don't forget about this (I have a backlog of things I have to resolve…).

For example, I don't like that the title of the section was renamed. The traps page (at least the sections that I wrote) are highly inspired by [BashPitfalls](https://mywiki.wooledge.org/BashPitfalls), and I highly recommend to steal ideas from that resource. For example, every title there is a code snippet, and that's very useful. You can just scroll through the titles to refresh your skills, without trying to figure out what titles like “Interpolation constructs” and such mean.

I'm not sure if the rewritten section conveys the “if you're using <{}>, then you're most likely wrong” idea, but it probably should.

threadless-screw commented:
@AlexDaniel Thanks. I now better understand what the idea behind the traps section is (love Wooledge's Bash site btw). I changed the title of the interpolation section, and removed some construct explanation (that can be found via the link) to put more emphasis on the code. I guess this is more in line with what you have in mind.

AlexDaniel commented:
I don't think I've seen anybody use `<$foo>`, but I did see people use `<{code}>` when they needed `$(code)`. Maybe it should be adjusted, but this looks good to me.

threadless-screw commented:
I think the discussed issues should have been resolved (but I may have screwed up with the pushed commits, not sure, because github does indeed display one of the issues as not outdated. Should the issue be present after a merge, I'll fix it.)
